### ISSUE TITLE
waitpid for Xwayland fork failed: No child processes

### ISSUE THREAD

**Issue by jbeich** [project-repo/cagebreak#23]
Similar to https://github.com/djpohly/dwl/issues/177. Xwayland works fine elsewhere: cage, hikari, kwinft, labwc, phoc, river, sway, wayfire.

```
$ cagebreak
[...]
00:00:00.085 [xwayland/server.c:92] Starting Xwayland on :0
00:00:00.116 [render/swapchain.c:105] Allocating new swapchain buffer
00:00:00.118 [render/allocator/gbm.c:142] Allocated 1280x720 GBM buffer (format 0x34325258, modifier 0x100000000000004)
00:00:00.118 [render/gles2/renderer.c:143] Created GL FBO for buffer 1280x720
00:00:00.160 [types/wlr_surface.c:748] New wlr_surface 0x880d81f80 (res 0x86f22d380)
00:00:00.165 [xwayland/server.c:243] waitpid for Xwayland fork failed: No child processes
(EE) failed to read Wayland events: Broken pipe
xterm: Xt error: Can't open display: :0

$ cc --version
FreeBSD clang version 11.0.1 (git@github.com:llvm/llvm-project.git llvmorg-11.0.1-0-g43ff75f2c3fe)
Target: x86_64-unknown-freebsd13.0
Thread model: posix
InstalledDir: /usr/bin

$ pkg info -x cagebreak wayland wlroots
cagebreak-1.8.3
wayland-1.20.0
wayland-protocols-1.25
xwayland-devel-21.0.99.1.171
wlroots-0.15.1
```

**Comment by project-repo** [comment]
Thanks for the bug report! I'm not quite sure what is causing this
atm... My best guess is that this is due to a version discrepancy
somewhere, seeing as the xwayland initialization code in
`cagebreak v1.8.3` and `cage v0.1.4` is pretty much identical as much as
I can tell. What version of `cage` are you using?
cheers
project-repo

**Comment by jbeich** [comment]
> What version of `cage` are you using? cheers project-repo

[Patched](https://github.com/freebsd/freebsd-ports/blob/55f41c0795d0/x11-wm/cage/Makefile#L7-L11) to fix some bugs + wlroots 0.15 compat:
- https://github.com/Hjdskes/cage/commit/70ecdcb02ea0
- https://github.com/Hjdskes/cage/commit/31a3571a7798
- https://github.com/Hjdskes/cage/commit/1a3ab3eb3ad0
- https://github.com/Hjdskes/cage/commit/ee748fec3adf

cage `v0.1.4-36-gf71844a` unmodified also works fine.

**Comment by jbeich** [comment]
Curiously, all compositors that "work fine" use `waitpid` with real PID instead of `WAIT_ANY` (-1).

**Comment by jbeich** [comment]
Nevermind, `waitpid` is probably fine but `sigaction` has slightly different semantics between BSD and Linux kernels.

**Comment by jbeich** [comment]
Nevermind again, I'm not familar with the API (don't write C much) to make any guesses.

**Comment by Sevz17** [comment]
See https://github.com/djpohly/dwl/issues/177#issuecomment-1081241033

**Comment by project-repo** [comment]
Thanks for reporting this @jbeich and for providing a solution @Sevz17!
This should be fixed in release 1.9.0 so I'm closing this for now. Feel
free to reopen if the bug persists though.
Cheers
project-repo
